charmcraft/commands/store/store.py

```diff
@@ -76,3 +88,29 @@ def list_registered_names(self):
         for item in response['charms']:
             result.append(Charm(name=item['name'], private=item['private'], status=item['status']))
         return result
+
+    def upload(self, name, filepath):
+        """Upload the content of filepath to the indicated charm."""
+        upload_id = self._client.push(filepath)
+
+        endpoint = '/v1/charm/{}/revisions'.format(name)
+        response = self._client.post(endpoint, {'upload-id': upload_id})
+        status_url = response['status-url']
+        logger.debug("Upload %s started, got status url %s", upload_id, status_url)
+
+        while True:
+            response = self._client.get(status_url)
+            logger.debug("Status checked: %s", response)
+
+            # as we're asking for a single upload_id, the response will always have only one item
+            (revision,) = response['revisions']
+            status = revision['status']
+
+            if status in UPLOAD_ENDING_STATUSES:
+                return Uploaded(
+                    ok=UPLOAD_ENDING_STATUSES[status],
+                    status=status, revision=revision['revision'])
+
+            # XXX Facundo 2020-06-30: Implement a slight backoff algorithm and fallout after
+            # N attempts (which should be big, as per snapcraft experience). Issue: #79.
```

Comment:
I feel like this should also eventually stop retrying. What happens if the status ends up as ERROR or some other string that isn't strictly 'rejected' or 'accepted' ?
If anything, I would think that we would only retry if we understand the status and it indicates we aren't done yet. (instead of always retrying unless we are sure we shouldn't)

Reply:
Yes, the #79 description involves a timeout.

Regarding on "what check", I talked with Matias (from the Store) and we agreed on check on those both tokens to flag "verification ended", and leave the rest for other possible intermediate verification steps (that may grow/shrink in time). 

I will ask for more certainties about this, anyway.